Content.Shared/Xenoarchaeology/Artifact/SharedXenoArtifactSystem.Graph.cs

```diff
@@ -0,0 +1,437 @@
+    {
+        if (ent.Comp.NodeVertices[index] is { } netUid && GetEntity(netUid) is var uid)
+            return (uid, XenoArtifactNode(uid));
+
+        throw new ArgumentException($"index {index} does not correspond to an existing node in {ToPrettyString(ent)}");
+    }
+
+    public bool TryGetNode(Entity<XenoArtifactComponent?> ent, int index, [NotNullWhen(true)] out Entity<XenoArtifactNodeComponent>? node)
+    {
+        node = null;
+        if (!Resolve(ent, ref ent.Comp))
+            return false;
+
+        if (index < 0 || index >= ent.Comp.NodeVertices.Length)
+            return false;
+
+        if (ent.Comp.NodeVertices[index] is { } netUid && GetEntity(netUid) is var uid)
+            node = (uid, XenoArtifactNode(uid));
+
+        return node != null;
+    }
+
+    /// <summary>
+    /// Gets the index of the first empty spot in the NodeVertices array.
+    /// If there is none, resizes both arrays and returns the new index.
+    /// </summary>
+    public int GetFreeNodeIndex(Entity<XenoArtifactComponent> ent)
+    {
+        var length = ent.Comp.NodeVertices.Length;
+        for (var i = 0; i < length; i++)
+        {
+            if (ent.Comp.NodeVertices[i] != null)
+                continue;
+
+            return i;
+        }
+
+        ResizeNodeGraph(ent, length + 1);
+        return length;
+    }
+
+    public IEnumerable<Entity<XenoArtifactNodeComponent>> GetAllNodes(Entity<XenoArtifactComponent> ent)
+    {
+        foreach (var netNode in ent.Comp.NodeVertices)
+        {
+            if (GetEntity(netNode) is { } node)
+                yield return (node, XenoArtifactNode(node));
+        }
+    }
+
+    public IEnumerable<int> GetAllNodeIndices(Entity<XenoArtifactComponent> ent)
+    {
+        for (var i = 0; i < ent.Comp.NodeVertices.Length; i++)
+        {
+            if (ent.Comp.NodeVertices[i] is not null)
+                yield return i;
+        }
+    }
+
+    public bool AddEdge(Entity<XenoArtifactComponent?> ent, EntityUid from, EntityUid to, bool dirty = true)
+    {
+        if (!Resolve(ent, ref ent.Comp))
+            return false;
+
+        if (!TryGetIndex(ent, from, out var fromIdx) ||
+            !TryGetIndex(ent, to, out var toIdx))
+            return false;
+
+        return AddEdge(ent, fromIdx.Value, toIdx.Value, dirty: dirty);
+    }
+
+    public bool AddEdge(Entity<XenoArtifactComponent?> ent, int fromIdx, int toIdx, bool dirty = true)
+    {
+        if (!Resolve(ent, ref ent.Comp))
+            return false;
+
+        DebugTools.Assert(fromIdx >= 0 && fromIdx < ent.Comp.NodeVertices.Length, $"fromIdx is out of bounds!");
+        DebugTools.Assert(toIdx >= 0 && toIdx < ent.Comp.NodeVertices.Length, $"toIdx is out of bounds!");
+
+        if (ent.Comp.NodeAdjacencyMatrix[fromIdx][toIdx])
+            return false; //Edge is already present
+
+        // TODO: add a safety check to prohibit cyclic paths.
```

Comment:
need to fix it before merging

Reply:
Maybe its okay? Current system prevents having cyclic graph by design of the way graph is initially created. I think this thing needs to be addressed with adding of code that will 'attach nodes' on the run - new devices that design doc was mentioning. They are not part of this PR